File: Day-6/CatchThieves.js
function maxThievesCaught(arr, k) {
  let police = [];
  let thieves = [];

  for (let i = 0; i < arr.length; i++) {
      if (arr[i] === 'P') police.push(i);
      else if (arr[i] === 'T') thieves.push(i);
  }

  let i = 0, j = 0, caught = 0;

  while (i < police.length && j < thieves.length) {
      let p = police[i];
      let t = thieves[j];

      if (Math.abs(p - t) <= k) {
          caught++;
          i++;
          j++;
      } else if (p < t) {
          i++;
      } else {
          j++;
      }
  }

  return caught;
}

let arr = ['P', 'T', 'T', 'P', 'T'];
let k = 1;
console.log("Maximum thieves caught:", maxThievesCaught(arr, k)); // Output: 2

File: Day-6/CatchThieves.go
package main

import (
	"fmt"
	"math"
)

func maxThievesCaught(arr []rune, k int) int {
	var police, thieves []int

	// Store positions of police and thieves
	for i, ch := range arr {
		if ch == 'P' {
			police = append(police, i)
		} else {
			thieves = append(thieves, i)
		}
	}

	i, j, caught := 0, 0, 0

	// Greedy matching
	for i < len(police) && j < len(thieves) {
		p := police[i]
		t := thieves[j]

		if math.Abs(float64(p-t)) <= float64(k) {
			caught++
			i++
			j++
		} else if p < t {
			i++
		} else {
			j++
		}
	}

	return caught
}

func main() {
	arr := []rune{'P', 'T', 'T', 'P', 'T'}
	k := 1
	fmt.Println("Maximum thieves caught:", maxThievesCaught(arr, k)) // Output: 2
}